### ISSUE TITLE
Can not read multipart content of as4 requests

### ISSUE THREAD

**Issue by yasharpm** [pondersource/peppol-php#133]
(context is I am sending from phase4 to peppolnext as4 end point)
Nextcloud's `Request` object is not capable to receive content of type `multipart/related`. I checked the class in the runtime and here is the source to it: [Request_on_nextcloud_source](https://github.com/nextcloud/server/blob/master/lib/private/AppFramework/Http/Request.php)
I tried to get the `inputStream` variable directly from the `Request` object and I am getting `null` which is weird. The variable is `protected` and from what I know from php so far, I should actually get an access exception not a null.
As the `inputStream` is never set to `null` in the code, I suspect the `Request` is being created with `null` as the `inputStream`? Which is also weird because I have double checked both the request method and the content length header and they are as expected.
Anyhow, I think we should either find a way to inject another implementation of the `IRequest` interface that supports `multipart/related` contents or use another library to receive the requests (I read somewhere in nextcloud documents that OCP packages are only recommended and not required).

**Comment by yasharpm** [comment]
Any idea how I am "not" getting access exception on the `inputStream`? I found [here](https://stackoverflow.com/questions/19589908/how-to-read-protected-variable) that there is a way to read the protected variables through reflection if need be.

**Comment by yasharpm** [comment]
We also have [this](https://www.example-code.com/phpExt/http_multipart_related_response.asp) from @navid-dada in case we need to deserialize the multi part body.

**Comment by yasharpm** [comment]
@michielbdejong Helpen!

**Comment by michielbdejong** [comment]
Probably the `inputStream` was already consumed by the time the Request object reaches the app code?
Can you give steps to reproduce? E.g. a `curl` command that POSTs a `multipart/related` body to PeppolNext's AS4 endpoint, and then show expected outputs and actual outputs so we can compare.
Try what happens if you use `PUT` as the verb and then call `$request->getContent()` in the controller. I think that should give you access to the raw request body.

**Comment by michielbdejong** [comment]
Or just change [this line](https://github.com/nextcloud/server/blob/master/lib/private/AppFramework/Http/Request.php#L400) from
> if ($this->method === 'PUT'

to
> if (true

**Comment by yasharpm** [comment]
> Probably the `inputStream` was already consumed by the time the Request object reaches the app code?

It is possible, but does it become null though? No where in code it is set to null. And as I said, I am expecting an access exception, not get a null.

> Can you give steps to reproduce?

Will do!

> Try what happens if you use `PUT` as the verb and then call `$request->getContent()` in the controller.

`$request->getContent()` actually returns an access exception since it is a `protected` method. But `$request->put` will do it. I'll try.

**Comment by yasharpm** [comment]
As discussed with @michielbdejong, I will create a pull request on the nextcloud repo to add the option to receive the raw body in case of a `POST` request with `multipart/related` header [here](https://github.com/nextcloud/server/blob/fc9a644bb54736721f14135be833f40bb8d6b124/lib/private/AppFramework/Http/Request.php#L414).

Next, I will use the library @navid-dada suggested to read the body inside our own code.

The `inputStream` null mystery is also solved. PHP returns `NULL` if you don't have access to a class variable. But throws an access Exception for calling inaccessible methods.

**Comment by michielbdejong** [comment]
See also https://github.com/pdsinterop/solid-nextcloud/blob/main/solid/lib/Controller/StorageController.php line 177

**Comment by yasharpm** [comment]
> did you try accessing $_POST or $_SERVER, actually? Maybe that would work too, to bypass Nextcloud's Request object.

I checked just now, they don't hold the body information. `$_POST` captures the submitted form data. i.e. `application/x-www-form-urlencoded`. `$_SERVER` and `$_REQUEST` had nothing useful as well.

 > See also https://github.com/pdsinterop/solid-nextcloud/blob/main/solid/lib/Controller/StorageController.php line 177

I could not understand the relation it has with the issue.

**Comment by michielbdejong** [comment]
Ah sorry! Line 277. I know that we can read the raw body of a POST in Solid-Nextcloud, regardless of the Content-Type.
https://github.com/pdsinterop/solid-nextcloud/blob/main/solid/lib/Controller/StorageController.php#L277 uses 

```
\Laminas\Diactoros\ServerRequestFactory::fromGlobals($_SERVER, $_GET, $_POST, $_COOKIE, $_FILES)
```

so the raw body must be in one of those parameters.

**Comment by yasharpm** [comment]
Seems like `file_get_contents('php://input')` will get the entire request body. And I just realized that the same command is used to take the request body in `Request.php`.
I made a change to capture the entire body of `multipart/related` POST requests and in the API end point used simple `explode` calls to read the body and it works. Problem is, some NextCloud tests are failing when I created a pull request. In a fresh setup, all the requests fail unless I comment out the change that I have made. But after the setup is complete, I uncomment it and everything is fine.
